independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/DeploymentInjectingDependencyVisitor.java

```diff
@@ -75,101 +95,275 @@ public DeploymentInjectingDependencyVisitor(MavenArtifactResolver resolver, List
+        final List<DependencyNode> deploymentDeps = deploymentNode.getChildren();
+        if (!replaceDirectDepBranch(extDep, deploymentDeps)) {
             throw new BootstrapDependencyProcessingException(
-                    "No dependencies collected for Quarkus extension deployment artifact " + newNode.getArtifact()
-                            + " while at least the corresponding runtime artifact " + originalNode.getArtifact()
-                            + " is expected");
+                    "Quarkus extension deployment artifact " + deploymentNode.getArtifact()
+                            + " does not appear to depend on the corresponding runtime artifact "
+                            + extDep.info.runtimeArtifact);
         }
-        log.debugf("Injecting deployment dependency %s", newNode);
 
-        originalNode.setData(QUARKUS_RUNTIME_ARTIFACT, originalNode.getArtifact());
-        originalNode.setArtifact(newNode.getArtifact());
-        originalNode.getDependency().setArtifact(newNode.getArtifact());
-        originalNode.setChildren(children);
+        final DependencyNode runtimeNode = extDep.runtimeNode;
+        runtimeNode.setData(QUARKUS_RUNTIME_ARTIFACT, runtimeNode.getArtifact());
+        runtimeNode.setArtifact(deploymentNode.getArtifact());
+        runtimeNode.getDependency().setArtifact(deploymentNode.getArtifact());
+        runtimeNode.setChildren(deploymentDeps);
+    }
+
+    private boolean replaceDirectDepBranch(ExtensionDependency extDep, List<DependencyNode> deploymentDeps)
+            throws BootstrapDependencyProcessingException {
+        int i = 0;
+        DependencyNode inserted = null;
+        while (i < deploymentDeps.size()) {
+            final Artifact a = deploymentDeps.get(i).getArtifact();
+            if (a == null) {
+                continue;
+            }
+            if (isSameKey(extDep.info.runtimeArtifact, a)) {
+                // we are not comparing the version in the above condition because the runtime version
+                // may appear to be different then the deployment one and that's ok
+                // e.g. the version of the runtime artifact could be managed by a BOM
+                // but overridden by the user in the project config. The way the deployment deps
+                // are resolved here, the deployment version of the runtime artifact will be the one from the BOM.
+                inserted = new DefaultDependencyNode(extDep.runtimeNode);
+                inserted.setChildren(extDep.runtimeNode.getChildren());
+                deploymentDeps.set(i, inserted);
+                break;
+            }
+            ++i;
+        }
+        if (inserted == null) {
+            return false;
+        }
+
+        if (extDep.runtimeExtensionDeps != null) {
+            for (ExtensionDependency dep : extDep.runtimeExtensionDeps) {
+                for (DependencyNode deploymentDep : deploymentDeps) {
+                    if (deploymentDep == inserted) {
```

Comment:
Is this deliberately an identity comparison?

Reply:
Yes, it is. It would mean the deploymentDep matches the just inserted node above.